tracer/src/Datadog.Trace/Configuration/ImmutableIntegrationSettingsCollection.cs

```diff
@@ -0,0 +1,104 @@
+namespace Datadog.Trace.Configuration
+{
+    /// <summary>
+    /// A collection of <see cref="ImmutableIntegrationSettings"/> instances, referenced by name.
+    /// </summary>
+    public class ImmutableIntegrationSettingsCollection
+    {
+        private readonly IConfigurationSource _source;
+        private readonly HashSet<string> _disabledIntegrations;
+        private readonly ConcurrentDictionary<string, ImmutableIntegrationSettings> _settingsByName;
+        private readonly Func<string, ImmutableIntegrationSettings> _valueFactory;
+        private readonly ImmutableIntegrationSettings[] _settingsById;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ImmutableIntegrationSettingsCollection"/> class.
+        /// </summary>
+        /// <param name="settings">The <see cref="IntegrationSettingsCollection"/> to populate the immutable settings.</param>
+        /// <param name="disabledIntegrationNames">Additional integrations that should be disabled</param>
+        public ImmutableIntegrationSettingsCollection(
+            IntegrationSettingsCollection settings,
+            HashSet<string> disabledIntegrationNames)
+        {
+            _source = settings.Source;
+            _disabledIntegrations = disabledIntegrationNames ?? throw new ArgumentNullException(nameof(disabledIntegrationNames));
+            _settingsById = GetIntegrationSettingsById(settings, disabledIntegrationNames);
+            _settingsByName = GetIntegrationSettingsByName(settings, disabledIntegrationNames);
+            _valueFactory = name =>
+            {
+                if (IntegrationRegistry.Ids.TryGetValue(name, out var id))
+                {
+                    return _settingsById[id];
+                }
+
+                // We have no id for this integration, it will only be available in _settingsByName
+                var initialSettings = new IntegrationSettings(name, _source);
+                var isExplicitlyDisabled = _disabledIntegrations.Contains(name);
+
+                return new ImmutableIntegrationSettings(initialSettings, isExplicitlyDisabled);
+            };
+        }
+
+        internal ICollection<string> DisabledIntegrations => _disabledIntegrations;
+
+        /// <summary>
+        /// Gets the <see cref="IntegrationSettings"/> for the specified integration.
+        /// </summary>
+        /// <param name="integrationName">The name of the integration.</param>
+        /// <returns>The integration-specific settings for the specified integration.</returns>
+        public ImmutableIntegrationSettings this[string integrationName] => this[new IntegrationInfo(integrationName)];
+
+        internal ImmutableIntegrationSettings this[IntegrationInfo integration]
+        {
+            get
+            {
+                return integration.Name == null
+                           ? _settingsById[integration.Id]
+                           : _settingsByName.GetOrAdd(integration.Name, _valueFactory);
```

Comment:
Doesn't seem very immutable.

Reply:
Seems to me this collection could be created in advance and made read-only, but not sure if the extra overhead is really worth it.

Reply:
> Seems to me this collection could be created in advance and made read-only, but not sure if the extra overhead is really worth it.

Yeah, this one was bugging me. Originally I was hoping to make it completely read-only, but a) we have to make the settings mutable, as we disable them at runtime if there's ducktyping errors, b) it's not clear to me why we need the dictionary _at all_. Given we only use _known_ integration IDs, `_settingsByName` should _always_ be empty as far as I can tell... But "Chesterton's fence" suggests that's not the case... so I kept it as-is for now. 

If someone else is happy to say that it's no longer necessary at all, then that makes things a _lot_ easier 😄

Reply:
This was meant to support looking up integrations by name if there was a version mismatch between `Datadog.Trace.dll` and `Datadog.Trace.ClrProfiler.Managed.dll` (i.e. if a new integration didn't exist in the older dll).
https://github.com/DataDog/dd-trace-dotnet/pull/1016#discussion_r519948941

Since this type of version mismatch can no longer happen, I don't think we need the string-based API at all anymore.

Reply:
I guess there's still a question here of whether any customers are using the existing `IntegrationSettingsCollection`, and if so, whether we want to support that, or if we should change the existing behaviour.

For example, customers _could_ do 

```csharp
var settings = TracerSettings.FromDefaultSources();
settings.Integrations["MyRandomIntegration"].Enabled = false;

if(settings.Integrations["MyRandomIntegration"].Enabled)
{
  // do some custom thing
}
```

Currently/previously this would add the `"MyRandomIntegration"` integration to the settings collection, which customers could then retrieve later. [Based on the above comment](https://github.com/DataDog/dd-trace-dotnet/pull/1951#discussion_r749437977), we will never use these, so the main question is how bad do we break customers. I see three options:

1. Make the indexer `internal`, so they _can't_ disable integrations in code this way. Not a great option.
2. Make the `string` indexer internal, make `IntegrationIds` public, and add an `IntegrationIds` indexer
3. Keep the indexer public, but if they request an "unknown" integration, throw and argument exception.
4. Keep the indexer public, but if they request an "unknown" integration just return a "default" value (which is effectively throw away). e.g. `new ImmutableIntegrationSettings(integration.Name)`. 

I'm leaning towards 3. Any other opinions?

Assuming we _do_ get rid of the string versions, that means we could also get rid of `IntegrationInfo` entirely (as `Name` is always `null`), which personally will make things a bit easier to follow I think (`IntegrationInfo` is really a discriminated union, which isn't obvious from the name!)

Reply:
Options 3 will break apps that are already using integration names we are removing in 2.0 (`AdoNet` and `ElasticsearchNet5`). This will happen at run time, not compile time, making it easy to happen in production. I think there's a higher change of that happening than your example:
```csharp
// this is wrong, settings.Integrations is meant for automatic instrumentation
if(settings.Integrations["MyRandomIntegration"].Enabled) { // do some custom thing }
```

I know we're already making many breaking changes, but the ones that can break a production app are scary. We should consider ignoring unknown integrations rather than throwing. Maybe log a warning message.

Reply:
There's two different scenarios here, depending whether we're talking about changes to `TracerSettings` or `ImmutableTracerSetings`. The following is _wrong_, but I agree, it should not throw or anything like that. A warning makes sense, as it's an incorrect usage of the API:

```csharp
var settings = TracerSettings.FromDefaultSources();

// this is wrong, settings.Integrations is meant for automatic instrumentation
if(settings.Integrations["MyRandomIntegration"].Enabled) { // do some custom thing }
```

In that case, for `TracerSettings.Integrations` we can still get rid of the dictionary in `IntegrationSettingsCollection`, log a warning when they try and access an invalid integration, and just return a new `IntegrationSettings` instance. The public API doesn't change.

The question is whether we do the _same_ for the `ImmutableTracerSettings`, e.g. 

```csharp
ImmutableTracerSettings settings = Tracer.Settings;

// this won't compile
settings.Integrations["MyRandomIntegration"].Enabled = false'

// this is wrong, settings.Integrations is meant for automatic instrumentation
if(settings.Integrations["MyRandomIntegration"].Enabled) { // do some custom thing }
```

Technically, as the `ImmutableTracerSettings` is a new API, we don't really have to expose the `Integrations` property at all. There's a few options:

* Make `Settings` on `Tracer` `internal`, and don't expose `ImmutableTracerSettings` at all. Probably the "easiest" fix, and shields us from API issues. But maybe customers want/need to do Tracer.Instance.Settings for some reason 🤷‍♂️ 
* Make the `Integrations` property on `ImmutableTracerSettings` `internal`. Means we don't have to expose the "string" based API at all on `Integrations` (and just use IntegrationIds`).
* Do the same as we do for `TracerSettings`, i.e. log an error, return "default" value

Reply:
Assuming we go with 3, this touches a lot of files, so I'd like to make the change in a follow up PR. The changes in this PR are "safe", in that they don't break any APIs etc. In the follow up I'll make them more-optimal by removing the dictionary etc